components/dashboard/sections/tax-information/TaxFormPreview.tsx

```diff
@@ -0,0 +1,106 @@
+
+import Image from '../../../Image';
+
+import { TaxFormType } from './common';
+
+const TaxFormLoadingPlaceholder = (
+  <Image src="/static/images/tax-form-placeholder.jpg" priority alt="" width={761} height={984} />
+);
+
+const PDFViewer = dynamic(() => import('../../../PDFViewer'), {
+  ssr: false,
+  loading: () => TaxFormLoadingPlaceholder,
+});
+
+const MemoizedTaxFormPreview = React.memo(({ url }: { url: string }) => {
+  const contentWrapperRef = React.useRef(null);
+  const [isInitialized, setIsInitialized] = React.useState(false);
+  const [currentRender, setCurrentRender] = React.useState(() => ({
+    url,
+    time: Date.now(),
+    result: (
+      <PDFViewer
+        loading={TaxFormLoadingPlaceholder}
+        pdfUrl={url}
+        contentWrapperRef={contentWrapperRef}
+        errorTextColor="neutral.900"
+        limitToPageWidth={false}
+        onLoadSuccess={() => setIsInitialized(true)}
+        onLoadError={() => setIsInitialized(true)}
+      />
+    ),
+  }));
+
+  const [nextRender, setNextRender] = React.useState(null);
+  const hasPendingReRender = currentRender.url !== url;
+  const debouncedSetNextRender = React.useMemo(() => debounce(setNextRender, 500), []);
+
+  // Pre-load the next render
+  React.useEffect(() => {
+    if (hasPendingReRender && url !== nextRender?.url) {
+      const newRender = { time: Date.now(), url, result: null };
+      newRender.result = (
+        <PDFViewer
+          pdfUrl={url}
+          loading={TaxFormLoadingPlaceholder}
+          contentWrapperRef={contentWrapperRef}
+          errorTextColor="neutral.900"
+          limitToPageWidth={false}
+          onLoadSuccess={() => {
+            setIsInitialized(true);
+            setCurrentRender(currentRender =>
+              currentRender.url !== newRender.url && currentRender.time < newRender.time ? newRender : currentRender,
+            );
+          }}
+          onLoadError={() => {
+            setIsInitialized(true);
+            setCurrentRender(currentRender =>
+              currentRender.url !== newRender.url && currentRender.time < newRender.time ? newRender : currentRender,
+            );
+          }}
+        />
+      );
+
+      debouncedSetNextRender(newRender);
+    }
+  }, [hasPendingReRender, debouncedSetNextRender, nextRender?.url, url]);
+
+  return (
+    <div ref={contentWrapperRef}>
+      <div key={currentRender.time} className={cn({ 'blur-sm': hasPendingReRender || !isInitialized })}>
+        {currentRender.result}
+      </div>
+      {hasPendingReRender && nextRender && nextRender.time !== currentRender.time && (
+        <div key={nextRender.time} className="hidden">
+          {nextRender.result}
+        </div>
+      )}
+    </div>
+  );
+});
+
+MemoizedTaxFormPreview.displayName = 'MemoizedTaxFormPreview';
+
+export const getTaxFormPreviewUrl = (type: TaxFormType, values, isFinal = false): string => {
+  const url = new URL(`${PDF_SERVICE_URL}/api/tax-form/${type}.pdf`);
```

Comment:
I was surprised by the `/api/` in the URL. Maybe this could be rewritten (as in "route rewrite") in the PDF service.